### PR TITLE
Generalise date calculation

### PR DESCRIPTION
**Pull request by dharmendrak**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RWA-2098

### Change description ###

Make date calculation generic

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

**Review by rasikasri — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/domain/entities/calendar/NextHearingDateIntervalData.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package uk.gov.hmcts.reform.wataskmanagementapi.domain.entities.calendar;
+
+import lombok.Builder;
+import lombok.Getter;
+
+import java.util.List;
+
+@Builder
+@Getter
+public class NextHearingDateIntervalData {
```
Comment (rasikasri) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/domain/entities/calendar/NextHearingDateIntervalData.java:10]: Why do you need a new class for next hearing date data and and due date data. You can use the DateTypeIntervalData class for all the types

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/DueDateIntervalReCalculator.java [hunk 2]
```diff
@@ -2,146 +2,106 @@
...
-                                         .map(CamundaValue::getValue)
-                                         .orElse(DUE_DATE_MUST_BE_WORKING_DAY_NEXT))
-            .dueDateTime(dueDateProperties.stream()
-                             .filter(r -> r.getName().getValue().equals(DUE_DATE_TIME))
-                             .filter(r -> r.getCanReconfigure().getValue().booleanValue() == TRUE)
-                             .reduce((a, b) -> b)
-                             .map(ConfigurationDmnEvaluationResponse::getValue)
-                             .map(CamundaValue::getValue)
-                             .orElse(DEFAULT_DUE_DATE_TIME))
+    private DateTypeIntervalData readDueDateOriginFields(List<ConfigurationDmnEvaluationResponse> dueDateProperties) {
```
Comment (rasikasri) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/DueDateIntervalReCalculator.java:44]: This method is same as the readDueDateOriginFields in the super class other than checking the canReconfigure flag. You can pass it as a parameter to the method. Please see how its done in this pr https://github.com/hmcts/wa-task-management-api/pull/915

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/DueDateCalculator.java [hunk 3]
```diff
@@ -2,47 +2,73 @@
...
-        var dueDateResponse = getProperty(dueDateProperties, DUE_DATE);
+    public ConfigurationDmnEvaluationResponse calculateDate(
+        List<ConfigurationDmnEvaluationResponse> dueDateProperties,
+        DateType dateType) {
+        var dueDateResponse = getProperty(dueDateProperties, DUE_DATE.getType());
         var dueDateTimeResponse = getProperty(dueDateProperties, DUE_DATE_TIME);
+        return calculatedDate(dateType, dueDateResponse, dueDateTimeResponse);
+    }
+
+    ConfigurationDmnEvaluationResponse calculatedDate(
```
Comment (rasikasri) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/DueDateCalculator.java:38]: This method should be protected visibility and rest of the methods here are private

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/DueDateTimeCalculator.java [hunk 4]
```diff
@@ -2,27 +2,46 @@
...
     @Override
-    public LocalDateTime calculateDueDate(List<ConfigurationDmnEvaluationResponse> dueDateProperties) {
-        var dueDateTimeResponse = getProperty(dueDateProperties, DUE_DATE_TIME);
-        return addTimeToDate(dueDateTimeResponse, DEFAULT_DATE);
+    public ConfigurationDmnEvaluationResponse calculateDate(List<ConfigurationDmnEvaluationResponse> dueDateProperties,
+                                                            DateType dateType) {
+        return calculatedDate(dateType, getProperty(dueDateProperties, DUE_DATE_TIME));
+    }
+
+    ConfigurationDmnEvaluationResponse calculatedDate(
```
Comment (rasikasri) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/DueDateTimeCalculator.java:37]: Same here, protected scope

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/NextHearingDateIntervalReCalculator.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    @Override
+    public ConfigurationDmnEvaluationResponse calculateDate(
+        List<ConfigurationDmnEvaluationResponse> nextHearingDateProperties,
+        DateType dateType) {
+        DateTypeIntervalData nextHearingDateIntervalData = readOriginFields(nextHearingDateProperties);
+
+        return calculateDate(dateType, nextHearingDateIntervalData);
+    }
+
+    private DateTypeIntervalData readOriginFields(List<ConfigurationDmnEvaluationResponse> dateTypeProperties) {
```
Comment (rasikasri) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/NextHearingDateIntervalReCalculator.java:52]: Same comment as of due date re-calculator, you can use the method in super class with canReconfigure parameter

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/PriorityDateIntervalReCalculator.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    @Override
+    public ConfigurationDmnEvaluationResponse calculateDate(
+        List<ConfigurationDmnEvaluationResponse> priorityDateProperties,
+        DateType dateType) {
+        return calculateDate(dateType, readPriorityDateOriginFields(priorityDateProperties));
+    }
+
+    private DateTypeIntervalData readPriorityDateOriginFields(
```
Comment (rasikasri) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/PriorityDateIntervalReCalculator.java:45]: same here

**Review by rasikasri — CHANGES_REQUESTED** [review]

**Review by rasikasri — DISMISSED** [review]

**Review by PrasadKumarP — APPROVED** [review]

**Pull request merged**